src/Groups/matrices/iso_nf_fq.jl

```diff
@@ -0,0 +1,172 @@
+# Detinko, Flannery, O'Brien "Recognizing finite matrix groups over infinite
+# fields", Section 4.2
+function isomorphic_group_over_finite_field(matrices::Vector{T}) where T <: Union{ fmpq_mat, AbstractAlgebra.MatElem{nf_elem} }
+   @assert !isempty(matrices)
+
+   # One should probably check whether all matrices are n by n (and invertible
+   # and such ...)
+
+   K = base_ring(matrices[1])
+   n = nrows(matrices[1])
+
+   Fq, matrices_Fq = good_reduction(matrices, 2)
+
+   G = MatrixGroup(n, Fq, matrices_Fq)
+   N = GAP.Globals.Order(G.X)
+   if maximal_order_of_finite_linear_group_qq(degree(K)*n) < N
+      error("Group is not finite")
+   end
+
+   G_to_fin_pres = GAP.Globals.IsomorphismFpGroupByGenerators(G.X, GAP.julia_to_gap([ g.X for g in gens(G) ]))
+   F = GAP.Globals.Range(G_to_fin_pres)
+   rels = GAP.Globals.RelatorsOfFpGroup(F)
+
+   inv_matrices = Vector{T}(undef, length(matrices))
+   inv_computed = falses(length(matrices))
+   for i = 1:length(rels)
+      rel = map(Int, GAP.Globals.LetterRepAssocWord(rels[i]))
+      M = identity_matrix(K, n)
+      for k in rel
+         if k < 0
+            k = -k
+            if !inv_computed[k]
+               inv_matrices[k] = inv(matrices[k])
+            end
+            M *= inv_matrices[k]
+         else
+            M *= matrices[k]
+         end
+      end
+      if !isone(M)
+         error("Group is not finite")
+      end
+   end
+   return G
```

Comment:
yeah, this really should return (also) a homomorphism

Reply:
I actually wasn't aware that we want to have the char 0 group also on the GAP side. But it probably will make things easier.

Reply:
Yes. The group does not have to be able to do much; GAP just needs to know how to multiple and invert elements. Basically it should work already now, except the Oscar matrix on the GAP side is not in the filter `IsMultiplicativeElementWithInverse` -- this is precisely what the glue code in `JuliaExperimental` does.

Well, I still did some manual experiments. One of the things I found is that there is no method `/(::fmpq_mat, ::fmpq_mat)` nor `//(::fmpq_mat, ::fmpq_mat)` and that breaks some GAP code which expects to be able to form quotients of matrices. But that is easily rectified.

```
julia> M = matrix(QQ, [ 0 1 0; -1 0 0; 0 0 -1 ])
[ 0   1    0]
[-1   0    0]
[ 0   0   -1]

julia> GAP.Globals.M = M
[ 0   1    0]
[-1   0    0]
[ 0   0   -1]

julia> GAP.Globals.m=GAP.julia_to_gap(M)
GAP: [ [ 0, 1, 0 ], [ -1, 0, 0 ], [ 0, 0, -1 ] ]

julia> GAP.prompt()
gap> M;
<Julia: [0 1 0; -1 0 0; 0 0 -1]>

gap> G := GroupByGenerators([M]);
<group with 1 generators>

gap> One(G);
<Julia: [1 0 0; 0 1 0; 0 0 1]>
gap> #Size(G);  # TODO: this does not work because a raw Julia matrix is not in the filter `IsMultiplicativeElementWithInverse`)
gap> H:=Group(m);
Group([ [ [ 0, 1, 0 ], [ -1, 0, 0 ], [ 0, 0, -1 ] ] ])
gap> Size(H);
4
gap> hom:=GroupHomomorphismByFunction(G, H, x -> JuliaToGAP(IsList, Julia.Matrix(x),true));
MappingByFunction( <group with 1 generators>, Group([ [ [ 0, 1, 0 ], [ -1, 0, 0 ], [ 0, 0, -1 ] ]
 ]), function( x ) ... end )

# now add two "missing" methods for left/right quotients
gap> InstallOtherMethod( \/, [ "IsJuliaObject", "IsJuliaObject" ], {a,b} -> a * b^-1);
gap> InstallOtherMethod( LeftQuotient, [ "IsJuliaObject", "IsJuliaObject" ], {a,b} -> a^-1 * b);

# now we can already do stuff
gap> IsInjective(hom);
true
gap> Kernel(hom);
<trivial group>
gap> IsSurjective(hom);
true
```

Reply:
And going further, we can now even use GAP's existing machinery which allows delegating computations through a monomorphism into a "nicer" group, even without fixing the `IsMultiplicativeElementWithInverse` filter bit, and without using `JuliaExperimental`:
```
gap> SetNiceMonomorphism(G, hom);
gap> SetIsHandledByNiceMonomorphism(G, true);
gap> Size(G);
4
```
Granted, this is a rather trivial example ;-)

Reply:
What is `//(::fmpq_mat, ::fmpq_mat)` supposed to do? Looks a bit ambiguous. Is it left or right divsion?

Reply:
a // b  should always be equal to a * inv(b), I see no ambiguity?

That said, I am not asking for such a method to be added... I mean, it might be useful for other reasons, but in the context here, our existing "adapter code" in GAP.jl would not call it anyway: it always delegates GAP's `/` to Julia's `/`, which is "right" in some cases and "wrong" in others; I could change it to always delegate to Julia's `//` but that would just make it right/wrong in other cases... (though perhaps it would be "right" in more cases that we care about? Dunno... Point is, that adapter code is only meant for quick&dirty hacks, for serious work, one needs to take the type of the Julia objects into account on the GAP side, and also possibly not work with "naked" Julia objects, but rather with GAP wrappers for them... Although it went surprisingly well in the example above :-)

Reply:
Hmm, I guess now I wonder why `/` is not supported for `fmpq_mat` and `fmpq`... It is supported for Julia rationals just fine:
```
julia> (1//2) / (3//5)
5//6

julia> a = [1//2 3//4 ;2//3 5//6]
2×2 Matrix{Rational{Int64}}:
 1//2  3//4
 2//3  5//6

julia> a / a
2×2 Matrix{Rational{Int64}}:
 1//1  0//1
 0//1  1//1

```

Reply:
Why not `inv(b) * a`?

Regarding `/`. Because `1/2 != (1//1)/(2//1)` so we kind of banned `/` everywhere.

Reply:
I don't understand: why does `1/2 != (1//1)/(2//1)` mean  that `/` should not be support for `fmpq_mat` and `fmpq` ? It is supported for `Rational{T}` after all?